### PR TITLE
feat: prettier-plugin-javascript-effects

### PR DESCRIPTION
**Pull request by cdaringe**
# problem

- no prettier support for effects

# solution

- add prettier support for effects
- document the hacky way i've done it
- format all the things
- turn on formatting for various things that were previously turned off
- re-orient the build to build, then _lint_ as we now rely on built plugin assets to execute the lint (prettier --check) step
- add a test that merely asserts that all fixtures parse and are formatted

### CODE REVIEW

File: packages/effects-runtime/examples/src/passingDataTest.js [hunk 1]
```diff
@@ -1,19 +1,19 @@
...
 };
 
-'use effects';
-try{
-    main();
-} handle default with (e){
-    if(e.type === 'timesTwo'){
-        timesTwoHandler(e.number);
-    }
+("use effects");
```
Comment (cdaringe) [inline packages/effects-runtime/examples/src/passingDataTest.js]: nooooo

File: packages/effects-runtime/examples/src/passingDataTest.js [hunk 2]
```diff
@@ -1,19 +1,19 @@
...
+try {
+  main();
+} handle default  with (e) {
```
Comment (cdaringe) [inline packages/effects-runtime/examples/src/passingDataTest.js:15]: wtf nooo! ok :face_with_head_bandage: i took a shortcut, will figure out how to do it right

File: test/__snapshots__/babel-plugin-transform-effects.spec.ts.snap [hunk 3]
```diff
@@ -605,7 +605,7 @@ module.exports.test = ({it, expect}) => {
     });
 
     it('Should throw if a parent calls an effect handled by a child function', async () => {
-        await expect(throwsUnhandledEffectError()).rejects.toThrowError("Encountered an unhandled effect :child")
+        await expect(throwsUnhandledEffectError()).rejects.toThrowError("Encountered an unhandled effect: child")
```
Comment (dpchamps) [inline test/__snapshots__/babel-plugin-transform-effects.spec.ts.snap]: 🤔 this shouldn't be like this

**Review by dpchamps — APPROVED** [review]
This is super exciting to see.

I suppose the only *right* way to do this would be for effects proposal to take off enough and then submit a PR to prettier 🧙‍♂️.

RE:

> We could fork prettier, apply our changes in our fork, and rewire their processes to publish only a subset of their code (just the js plugin), and distribute that way. We should probably so something like that soon, so as to dedupe distribution of our @babel/parser.

What was your reasoning behind _not_ choosing this right off the bat? I honestly don't see too huge of a deal doing it the way you've implemented it now. I mean, we've extended the ECMAScript specification hahaha there's bound to be not-so-pretty things we gotta do for interop with other libs.

File: packages/effects-runtime/src/runtime.ts [hunk 4]
```diff
@@ -22,7 +22,7 @@ export class UnhandledEffectError extends TypeError {
   constructor({ type }: Effect) {
     super();
 
-    this.message = `Encountered an unhandled effect :${String(type)}`;
```
Comment (dpchamps) [inline packages/effects-runtime/src/runtime.ts]: This spacing was intentional, `:<effect-name>` in an attempt to try to communicate that the effect type was sort of meta, like a Symbol.

**Review by dpchamps — CHANGES_REQUESTED** [review]

File: packages/effects-runtime/src/runtime.ts (hunk 4 above)
Comment (cdaringe) [reply packages/effects-runtime/src/runtime.ts]: ya, i got the above error, and i thought "oh no, i've fat fingered an extra character somewhere" and then proceeded to search the codebase for `:childPerform`. `:childPerform` (obviously) didn't show up and i scratched my head. i then opened up my search for the error message prefix and saw that this statement was doing the thing. i get the appeal, and the desire for a cool ux here. it had the opposite initial effect for me, but maybe i was just being a dummy. i'll swap it back when i pick back up here. food for thought.

Comment (dpchamps) [reply packages/effects-runtime/src/runtime.ts]: It _may_ just be too cute. There's no realy rhyme or reason for this tag-style syntax... I say we remove if it was even the slightest confusing.

**Comment by cdaringe** [comment]
> What was your reasoning behind not choosing this right off the bat?

ya, fair question. i intended this new package's source layout to look like a plugin, not a top-level executable. thus, forking was never even considered until after all of the work was done and i was opening the PR. after discovering that i could use the effects parser 👌, things were looking good--initial experiments only failed at the print phase. i figured, "well, if it's only the printer" maybe i could recycle & copy prettier internals somewhat painlessly, mutate the copy for effects, and pass stuff back in via the plugin api. and in fact that's still more or less what i did, whilst incurring two costs: 1) literally coping prettier src files (they are not extensible), and 2) build/bundle those copied files post-effects-annotated-mutation. then i realized that without repeating that process for the `parser-babel.js` file too, we couldn't exploit all of the babel init sugar and TS compat stuff--so i ended up building _two files_. edits are probably < 15 LOC. cheap. the outstanding question is, what are the cost differences between what i've done vs forking, rewiring that fork drastically enough so that it's a plugin, making it s.t. upstream updates can still be applied meaningfully, and ensuring clean integration with the rest of the effects project (mainly the parser). maybe it'd be better. i dunno. after typing all of this out, i'm actually _less convinced_. something to play with

edit.

i typed that all out on my phone and it's prettty much incoherent 😆 . let's talk in person!

File: test/__snapshots__/babel-plugin-transform-effects.spec.ts.snap (hunk 3 above)
Comment (cdaringe) [reply test/__snapshots__/babel-plugin-transform-effects.spec.ts.snap]: e3c4cdf

File: packages/effects-runtime/src/runtime.ts (hunk 4 above)
Comment (cdaringe) [reply packages/effects-runtime/src/runtime.ts]: e3c4cdf

**Comment by cdaringe** [comment]
bump. this is ready

**Comment by cdaringe** [comment]
_double bump_

**Review by dpchamps — APPROVED** [review]

**Pull request merged**